sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -274,24 +278,32 @@ case class InSet(child: Expression, hset: Set[Any]) extends UnaryExpression with
     }
   }
 
-  def getHSet(): Set[Any] = hset
+  @transient private[this] lazy val set = child.dataType match {
+    case _: StructType =>
```

Comment:
I checked and indeed it is needed for ArrayType too. MapType is not comparable.
I think dealing with uncomparable types should be different issue. It wouldn't work before either. In.doCodeGen would fail with IllegalArgumentException. EqualTo has a check specifically for Map to throw AnalysisException. That could be a fix. Or implement comparing for Maps.